src/object.d

```diff
@@ -3722,6 +3722,186 @@ template _arrayOp(Args...)
     alias _arrayOp = arrayOp!Args;
 }
 
+// switch template - replaces code in rt/switch_.d
+
+// switch template - replaces code in rt/switch_.d
+int __switch(T)(T[] condition, T[][] caseLabels...)
+in
+{
+    assert(caseLabels.length >= 0);
+    assert(condition.length >= 0);
+
+    // Make sure caseLabels[] is sorted correctly
+
+    // Linear search not accepcaseLabels, but we iterate through the whole
+    // caseLabels anyway to make sure it's sorted so that we can use the binary search.
+    for (size_t j = 1u; j < caseLabels.length; j++)
+    {
+        auto len1 = caseLabels[j - 1].length;
+        auto len2 = caseLabels[j].length;
+
+        assert(len1 <= len2);
+        if (len1 == len2)
+        {
+            int ci;
+
+            ci = __cmp(caseLabels[j - 1], caseLabels[j]);
+            assert(ci < 0); // ci==0 means a duplicate
+        }
+    }
+}
+out (result)
+{
+    int cj;
+
+    if (result == -1)
+    {
+        // Not found
+        for (auto i = 0u; i < caseLabels.length; i++)
+        {
+            if (caseLabels[i].length == condition.length)
+            {   cj = __cmp(caseLabels[i], condition);
+                assert(cj != 0);
+            }
+        }
+    }
+    else
+    {
+        assert(0 <= result && cast(size_t)result < caseLabels.length);
+        for (auto i = 0u; 1; i++)
+        {
+            assert(i < caseLabels.length);
+            if (caseLabels[i].length == condition.length)
+            {
+                cj = __cmp(caseLabels[i], condition);
+                if (cj == 0)
+                {
+                    assert(i == result);
+                    break;
+                }
+            }
+        }
+    }
+}
+body
+{
+    size_t low  = 0;
+    size_t high = caseLabels.length;
+
+    if (high &&
+        condition.length >= caseLabels[0].length &&
+        condition.length <= caseLabels[high - 1].length)
+    {
+        // Looking for 0 length string, which would only be at the beginning
+        if (condition.length == 0)
+            return 0;
+
+        T c1 = condition[0];
+
+        // Do binary search
+        while (low < high)
+        {
+            auto mid = (low + high) >> 1;
+            auto pca = caseLabels[mid];
+            auto c   = cast(sizediff_t)(condition.length - pca.length);
+            if (c == 0)
+            {
+                c = cast(ubyte)c1 - cast(ubyte)pca[0];
+                if (c == 0)
+                {
+                    c = __cmp(condition, pca);
+                    if (c == 0)
+                    {
+                        return cast(int)mid;
+                    }
+                }
+            }
+            if (c < 0)
+            {
+                high = mid;
+            }
+            else
+            {
+                low = mid + 1;
+            }
+        }
+    }
```

Comment:
This is longer than the canonical implementation. Please rewrite with an eye for efficiency, simplicity, and clarity. FWIW all you need is:

```d
size_t low = 0, high = caseLabels.length;
while (low < high)
{
    auto mid = (low + high) / 2;
    auto r = __cmp(condition, caseLabels[mid]);
    if (r == 0) return cast(int) mid;
    if (r > 0) low = mid + 1;
    else high = mid;
}
return -1;
```

Since we know `caseLabels.length > 0`, we can save one comparison:

```d
size_t low = 0, high = caseLabels.length;
do
{
    auto mid = (low + high) / 2;
    auto r = __cmp(condition, caseLabels[mid]);
    if (r == 0) return cast(int) mid;
    if (r > 0) low = mid + 1;
    else high = mid;
}
while (low < high);
return -1;
```